packages/theme-ui/src/custom-properties.js

```diff
@@ -0,0 +1,97 @@
+// convert theme and style objects to CSS custom properties
+import { css } from '@styled-system/css'
+
+const toVarName = key => `--theme-ui-${key}`
+const toVarValue = (key, value) => `var(${toVarName(key)}, ${value})`
+
+const join = (...args) => args.filter(Boolean).join('-')
+
+const numberScales = {
+  fontWeights: true,
+  lineHeights: true,
+}
+const reservedKeys = {
+  useCustomProperties: true,
+  initialColorMode: true,
+}
+
+const toPixel = (key, value) => {
+  if (typeof value !== 'number') return value
+  if (numberScales[key]) return value
+  return value + 'px'
+}
+
+// convert theme values to custom properties
+export const toCustomProperties = (obj, parent, themeKey) => {
+  const next = Array.isArray(obj) ? [] : {}
+
+  for (let key in obj) {
+    const value = obj[key]
+    const name = join(parent, key)
+    if (value && typeof value === 'object') {
+      next[key] = toCustomProperties(value, name, key)
+      continue
+    }
+    if (reservedKeys[key]) {
+      next[key] = value
+      continue
+    }
+    const val = toPixel(themeKey || key, value)
+    next[key] = toVarValue(name, val)
+  }
+
+  return next
+}
+
+export const objectToVars = (parent, obj) => {
+  let vars = {}
+  for (let key in obj) {
+    if (key === 'modes') continue
+    const name = join(parent, key)
+    const value = obj[key]
+    if (value && typeof value === 'object') {
+      vars = {
+        ...vars,
+        ...objectToVars(name, value),
+      }
+    } else {
+      vars[toVarName(name)] = value
+    }
+  }
+  return vars
+}
+
+// create body styles for color modes
+export const createColorStyles = theme => {
+  if (!theme.colors || !theme.colors.modes) return {}
+  const { modes } = theme.colors
+  const styles = theme.useCustomProperties
+    ? objectToVars('colors', theme.colors)
+    : {}
+
+  Object.keys(modes).forEach(mode => {
+    const key = `&.theme-ui-${mode}`
+    if (!theme.useCustomProperties) {
+      styles[key] = {
+        color: modes[mode].text,
+        bg: modes[mode].background,
+      }
+    } else {
+      styles[key] = objectToVars('colors', modes[mode])
+    }
+  })
+
+  if (!theme.useCustomProperties) {
```

Comment:
The logic here could be simplified if we only support the FOUC issue in browser that support custom properties, which I suspect is an okay tradeoff

Reply:
Yeah, I'd say it's fine to have FOUC for legacy browsers since it will fall back to a small flash on initial load but otherwise work fine.